editor/modes/visual-editor/block.js

```diff
@@ -11,84 +11,116 @@ import Toolbar from 'components/toolbar';
 import BlockMover from 'components/block-mover';
 import BlockSwitcher from 'components/block-switcher';
 
-function VisualEditorBlock( props ) {
-	const { block } = props;
-	const settings = wp.blocks.getBlockSettings( block.blockType );
+class VisualEditorBlock extends wp.element.Component {
+	constructor() {
+		super( ...arguments );
+		this.bindBlockNode = this.bindBlockNode.bind( this );
+	}
 
-	let BlockEdit;
-	if ( settings ) {
-		BlockEdit = settings.edit || settings.save;
+	bindBlockNode( node ) {
+		this.node = node;
 	}
 
-	if ( ! BlockEdit ) {
-		return null;
+	componentWillReceiveProps( newProps ) {
+		if (
+			this.props.order !== newProps.order &&
+			this.props.isSelected &&
+			newProps.isSelected
+		) {
+			this.previousOffset = this.node.getBoundingClientRect().top;
+		}
 	}
 
-	const { isHovered } = props;
-	const isSelected = props.isSelected;
-	const className = classnames( 'editor-visual-editor__block', {
-		'is-selected': isSelected,
-		'is-hovered': isHovered
-	} );
+	componentDidUpdate() {
+		if ( this.previousOffset ) {
+			window.scrollTo(
+				window.scrollX,
+				window.scrollY + this.node.getBoundingClientRect().top - this.previousOffset
+			);
+			this.previousOffset = false;
+		}
+	}
 
-	const { onChange, onSelect, onDeselect, onMouseEnter, onMouseLeave, onInsertAfter } = props;
+	render() {
+		const { block } = this.props;
+		const settings = wp.blocks.getBlockSettings( block.blockType );
 
-	function setAttributes( attributes ) {
-		onChange( {
-			attributes: {
-				...block.attributes,
-				...attributes
-			}
+		let BlockEdit;
+		if ( settings ) {
+			BlockEdit = settings.edit || settings.save;
+		}
+
+		if ( ! BlockEdit ) {
+			return null;
+		}
+
+		const { isHovered, isSelected } = this.props;
+		const className = classnames( 'editor-visual-editor__block', {
+			'is-selected': isSelected,
+			'is-hovered': isHovered
 		} );
-	}
 
-	function maybeDeselect( event ) {
-		// Annoyingly React does not support focusOut and we're forced to check
-		// related target to ensure it's not a child when blur fires.
-		if ( ! event.currentTarget.contains( event.relatedTarget ) ) {
-			onDeselect();
+		const { onChange, onSelect, onDeselect, onMouseEnter, onMouseLeave, onInsertAfter } = this.props;
+
+		function setAttributes( attributes ) {
```

Comment:
If we're making this a proper class, these might ought to be proper member functions now.